code/drasil-lang/Language/Drasil/Space.hs

```diff
@@ -25,9 +25,20 @@ data Space =
   | Actor String
   | DiscreteD [Double]
   | DiscreteS [String] --ex. let Meal = {"breakfast", "lunch", "dinner"}
+  | Mapping [Space] Space -- TODO: this one shouldn't be exposed directly, only through a smart constructor
   | Void
   deriving (Eq, Show)
 
+mkPrimitiveMapping :: [Space] -> Space -> Space
+mkPrimitiveMapping _   (Mapping _ _) = error "Target must be basic"
+mkPrimitiveMapping ins trg
+  | any isMapping ins                = error "All inputs must be basic"
+  | otherwise                        = Mapping ins trg
+
+isMapping :: Space -> Bool
```

Comment:
`isMapping` is probably an anti-pattern (see 'boolean blindness'), so you don't want to use it.